File: golang/_0022_Generate_Parentheses/main.go
package main

func generateParenthesis(n int) []string {
	res := make([]string, 0)
	if n > 0 {
		dfs(n, 0, 0, &res, "")
	}

	return res
}

func dfs(n, left, right int, res *[]string, curRes string) {
	if left > n {
		return
	} else if right > left {
		return
	} else if left == n && right == n {
		*res = append(*res, curRes)
		return
	} else {
		dfs(n, left+1, right, res, curRes+"(")
		dfs(n, left, right+1, res, curRes+")")
	}
}

File: python3/_0022_Generate_Parentheses/main.py
from typing import List


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        if not n:
            return []

        res = []
        self.dfs(res, "", n, n)
        return res

    def dfs(self, res: List[str], oneRes: str, left: int, right: int):
        if not left and not right:
            res.append(oneRes)
            return
        elif left > right or left < 0:
            return

        self.dfs(res, oneRes + '(', left - 1, right)
        self.dfs(res, oneRes + ')', left, right - 1)
